openvm/src/utils.rs

```diff
@@ -13,17 +14,53 @@ use openvm_stark_backend::{
     interaction::Interaction,
     p3_field::PrimeField32,
 };
-use powdr_autoprecompiles::legacy_expression::{
-    AlgebraicExpression, AlgebraicReference, PolyID, PolynomialType,
-};
+use powdr_autoprecompiles::legacy_expression::{AlgebraicReference, PolyID, PolynomialType};
+use powdr_expression::AlgebraicExpression;
 use powdr_expression::{
     AlgebraicBinaryOperation, AlgebraicBinaryOperator, AlgebraicUnaryOperation,
     AlgebraicUnaryOperator,
 };
 use powdr_number::{BabyBearField, FieldElement};
 
+pub enum OpenVmReference {
+    /// Reference to a witness column. The boolean indicates if the reference is to the next row.
+    WitnessColumn(AlgebraicReference, bool),
+    IsFirstRow,
+    IsLastRow,
+    IsTransition,
+}
```

Comment:
I find it odd that we can't we access `first_row'`. Shouldn't columns and offsets be orthogonal?

I had a proposal ages ago which here would look like:
```
enum Expression {
   Reference(ShiftedReference)
}

struct ShiftedReference {
   reference: Reference,
   shift: bool // or an number for larger offsets
}

enum Reference {
   FirstRow,
   Witness(WitnessReference)
}
```